File: blas/cpp/hfloat64_arm64.cpp
#include <arm_neon.h>

void hsum_float64(double *input, double *result, uint64_t size) {
    *result = 0.0;
    int chunks = size / 2;

    for (int i = 0; i < chunks; i++) {
        *result += vaddvq_f64(((float64x2_t *) input)[i]);
    }
    // Leftovers
    for (int i = chunks * 2; i < size; i++) {
        *result += input[i];
    }
}

void hmax_float64(double *input, double *result, uint64_t size) {
    *result = input[0];
    int chunks = size / 2;

    for (int i = 0; i < chunks; i++) {
        double tmp = vmaxvq_f64(((float64x2_t *) input)[i]);
        if (tmp > *result) {
            *result = tmp;
        }
    }
    // Leftovers
    for (int i = chunks * 2; i < size; i++) {
        if (input[i] > *result) {
            *result = input[i];
        }
    }
}

void hmin_float64(double *input, double *result, uint64_t size) {
    *result = input[0];
    int chunks = size / 2;

    for (int i = 0; i < chunks; i++) {
        double tmp = vminvq_f64(((float64x2_t *) input)[i]);
        if (tmp < *result) {
            *result = tmp;
        }
    }
    // Leftovers
    for (int i = chunks * 2; i < size; i++) {
        if (input[i] < *result) {
            *result = input[i];
        }
    }
}

File: blas/hfloat64_arm64.go
package blas

import (
	"unsafe"
)

//go:noescape

func _hsum_float64(input, result unsafe.Pointer, size uint64)
func HsumFloat64(input []float64) float64 {
	var sum float64
	_hsum_float64(unsafe.Pointer(&input[0]), unsafe.Pointer(&sum), uint64(len(input)))
	return sum
}

func _hmax_float64(input, result unsafe.Pointer, size uint64)
func HmaxFloat64(input []float64) float64 {
	var max float64
	_hmax_float64(unsafe.Pointer(&input[0]), unsafe.Pointer(&max), uint64(len(input)))
	return max
}

func _hmin_float64(input, result unsafe.Pointer, size uint64)
func HminFloat64(input []float64) float64 {
	var min float64
	_hmin_float64(unsafe.Pointer(&input[0]), unsafe.Pointer(&min), uint64(len(input)))
	return min
}
